### DISCUSSION TITLE
Can someone explain ContentQueryAsync & QueryListItemsAsync

### DISCUSSION THREAD

**Discussion by daene17** [OrchardCMS/OrchardCore#9605] [Q&A]
So I am having trouble getting content query async to work. 

Here is my SQL query which I have tested and works in orchard 

![image](https://user-images.githubusercontent.com/8889105/119854985-66a7b100-bec6-11eb-8bf0-c92f5d72f3af.png)

I have tried to use it on a Razor page like so: 

`ContentQueryAsync("SELECT * FROM [ContentItemIndex] WHERE [ContentType] = 'Listing';"); `
but no dice. I also tried using the Query Name, that also did not work. So how do I get this to give me results? 

**Part 2 of this Question:** Can someone please explain to me in depth how 
`QueryListItemsAsync(string listContentItemId, Expression<Func<ContentItemIndex, bool>> itemPredicate = null)`
is used? The documentation is not very detailed.

**Comment by sebastienros** [comment]
`ContentQueryAsync` takes the name of the query, not the SQL: https://docs.orchardcore.net/en/dev/docs/reference/modules/Queries/#razor-helpers

So if you created the query like in your screenshot, it must have a name, use it instead.

`QueryListItemsAsync` is using the content item id of a List (for instance a Blog is a List) and returns all the content items that are in this list.

**Comment by Skrypt** [comment]
"SELECT DocumentId FROM ContentItemIndex" to get the "id".

**Comment by daene17** [comment]
@sebastienros That's not working, lets say I have this query named QueryAll:
![image](https://user-images.githubusercontent.com/8889105/119906597-5a8e1480-bf03-11eb-855c-3b2f35781e77.png)

Would I just use: `var listOfContentItems = Orchard.QueryAsync("QueryAll");`  ?

Because that's not working

**Comment by Skrypt** [comment]
Look at the comment under the checkbox that states that when checked, the result has to be a list of document ids.
So, your SQL Query is wrong in that case. You need to do "SELECT DocumentId FROM ContentItemIndex" so that it returns the documents corresponding to these elements from the ContentItemIndex table.

**Comment by daene17** [comment]
> Look at the comment under the checkbox that states that when checked, the result has to be a list of document ids.
> So, your SQL Query is wrong in that case. You need to do "SELECT DocumentId FROM ContentItemIndex" so that it returns the documents corresponding to these elements from the ContentItemIndex table.

I'm not sure I understand. Where is my SQL Query wrong? It is returning results in the actual cms. 

![image](https://user-images.githubusercontent.com/8889105/120546396-a7129d80-c3a4-11eb-8f00-de8261d8c109.png)

**Comment by Skrypt** [comment]
That's the record from the ContentItemIndex not the Document table. When you want to have all the fields from a Content Item you get the data from the Document.Content table. Here the checkbox does that for you when you want to returns "DOCUMENTS". It does a "SELECT * FROM Document WHERE Id IN ( the id's from your query )". The data in the ContentItemIndex is just an indexation of the data contained in the Document.Content table. So it contains less data.

By the way, your screenshot doesn't show if you checked the checkbox or not so I can't tell what you are doing exactly.
I think this should be converted to a discussion as there is no issue here.

**Reply by daene17** [reply]
> That's the record from the ContentItemIndex not the Document table. When you want to have all the fields from a Content Item you get the data from the Document.Content table. Here the checkbox does that for you when you want to returns "DOCUMENTS". It does a "SELECT * FROM Document WHERE Id IN ( the id's from your query )". The data in the ContentItemIndex is just an indexation of the data contained in the Document.Content table. So it contains less data.
> 
> By the way, your screenshot doesn't show if you checked the checkbox or not so I can't tell what you are doing exactly.
> I think this should be converted to a discussion as there is no issue here.

Okay I see. So now let's say with the following settings:
![image](https://user-images.githubusercontent.com/8889105/120704766-db04c600-c46b-11eb-875b-93d8371f9e69.png)
![image](https://user-images.githubusercontent.com/8889105/120704852-f4a60d80-c46b-11eb-8d38-3c643532a66b.png)

What would I put inside of this method: `var listOfContentItems = Orchard.QueryAsync("");`
to return a list of objects?

That's all I want to know is what do I need to pass to get results. I am sorry if I keep asking the same question, but I've tried multiple ways and it still keeps coming up null.

**Reply by Skrypt** [reply]
Orchard.QueryAsync("NameOfYourQuery")
